ludwig/data/dataset/tfrecord.py

```diff
@@ -0,0 +1,211 @@
+from ludwig.utils.misc_utils import get_combined_features, get_proc_features
+
+
+class TFRecordDataset(Dataset):
+    def __init__(self, url, features, training_set_metadata):
+        self.url = to_url(url)[7:]
+        if not os.path.isdir(self.url):
+            raise RuntimeError("url for TFRecordDataset must be a folder.")
+        abs_path = os.path.abspath(os.path.expanduser(self.url))
+        self.file_names = [os.path.join(abs_path, name)
+                           for name in glob.glob(os.path.join(abs_path, "*.tfrecords.gz"))]
+        self.features = [feature[PROC_COLUMN] for feature in features]
+        self.training_set_metadata = training_set_metadata
+
+        # dataset = tf.data.TFRecordDataset(self.file_names, compression_type="GZIP")
+        # self.feature, self.feature_lists = self._detect_schema(dataset)
+        with open(os.path.join(abs_path, "meta.json")) as in_file:
+            meta = json.load(in_file)
+        self.size = meta["size"]
+
+        self.reshape_features = {
+            feature[PROC_COLUMN]: list((-1, *training_set_metadata[feature[NAME]]['reshape']))
+            for feature in features
+            if "reshape" in training_set_metadata[feature[NAME]]
+        }
+
+    def get(self, feature_name, sample):
+        t = sample[feature_name]
+        reshape_dim = self.reshape_features.get(feature_name)
+        if reshape_dim is not None:
+            # When we read a 1D array from disk, we need to reshape it back to its
+            # full dimensions.
+            t = tf.reshape(t, reshape_dim)
+        return t
+
+    def __len__(self):
+        return self.size
+
+    @contextlib.contextmanager
+    def initialize_batcher(self,
+                           batch_size=128,
+                           should_shuffle=True,
+                           seed=0,
+                           ignore_last=False,
+                           horovod=None):
+        cur_shard, shard_count = None, None
+        if horovod:
+            cur_shard, shard_count = horovod.rank(), horovod.size()
+
+        dataset = tf.data.TFRecordDataset(self.file_names, compression_type="GZIP")
+        if shard_count > 1:
+            dataset = dataset.shard(shard_count, cur_shard)
+        total_samples = self.size
+        local_samples = int(total_samples / shard_count) if shard_count else total_samples
+
+        # map parser
+        features, feature_lists = self._detect_schema(dataset)
+        parser = self._get_parser(features, feature_lists)
+        dataset = dataset.map(parser)
+        # TODO(Hao): figure out the comments below.
+        # dataset = dataset.unbatch()
+        if should_shuffle:
+            # rows_per_piece = max([piece.get_metadata().num_rows for piece in reader.dataset.pieces])
+            # buffer_size = min(rows_per_piece, local_samples)
+            buffer_size = local_samples
+            dataset = dataset.shuffle(buffer_size)
+        dataset = dataset.repeat()
```

Comment:
I have to repeat the data set forever to avoid `EndOfIterator` issue. See [here](https://github.com/zhisbug/ludwig/blob/hao-add-tfrecord/ludwig/data/dataset/tfrecord.py#L82). I suppose this `dataset.repeat()` was implicitly called inside the `make_petastorm_dataset` [call ](make_petastorm_dataset)?

Reply:
@tgaddair  confirms fine!